skimage/filters/thresholding.py

```diff
@@ -268,46 +318,54 @@ def threshold_otsu(image, nbins=256):
     -----
     The input image must be grayscale.
     """
-    if image.ndim > 2 and image.shape[-1] in (3, 4):
+    if image is not None and image.ndim > 2 and image.shape[-1] in (3, 4):
         msg = "threshold_otsu is expected to work correctly only for " \
               "grayscale images; image shape {0} looks like an RGB image"
         warn(msg.format(image.shape))
 
-    # Check if the image is multi-colored or not
-    first_pixel = image.ravel()[0]
-    if np.all(image == first_pixel):
-        return first_pixel
+    # Check if the image has more than one intensity value; if not, return that
+    # value
+    if image is not None:
+        first_pixel = image.ravel()[0]
+        if np.all(image == first_pixel):
+            return first_pixel
 
-    hist, bin_centers = histogram(image.ravel(), nbins, source_range='image')
-    hist = hist.astype(float)
+    counts, bin_centers = _validate_image_histogram(image, hist, nbins)
 
     # class probabilities for all possible thresholds
-    weight1 = np.cumsum(hist)
-    weight2 = np.cumsum(hist[::-1])[::-1]
+    weight1 = np.cumsum(counts)
+    weight2 = np.cumsum(counts[::-1])[::-1]
     # class means for all possible thresholds
-    mean1 = np.cumsum(hist * bin_centers) / weight1
-    mean2 = (np.cumsum((hist * bin_centers)[::-1]) / weight2[::-1])[::-1]
+    mean1 = np.cumsum(counts * bin_centers) / weight1
+    mean2 = (np.cumsum((counts * bin_centers)[::-1]) / weight2[::-1])[::-1]
 
     # Clip ends to align class 1 and class 2 variables:
     # The last value of ``weight1``/``mean1`` should pair with zero values in
     # ``weight2``/``mean2``, which do not exist.
     variance12 = weight1[:-1] * weight2[1:] * (mean1[:-1] - mean2[1:]) ** 2
 
     idx = np.argmax(variance12)
-    threshold = bin_centers[:-1][idx]
+    threshold = bin_centers[idx]
+
     return threshold
 
 
-def threshold_yen(image, nbins=256):
+def threshold_yen(image=None, nbins=256, *, hist=None):
     """Return threshold value based on Yen's method.
+    Either image or hist must be provided. In case hist is given, the actual
+    histogram of the image is ignored.
 
     Parameters
     ----------
-    image : (N, M) ndarray
+    image : (N, M) ndarray, optional
         Input image.
     nbins : int, optional
         Number of bins used to calculate histogram. This value is ignored for
         integer arrays.
+    hist : array, or 2-tuple of arrays, optional
+        Histogram to determine the threshold from and a corresponding array
```

Comment:
I would make the same suggestion as above. Also, I wonder if 'Alternative' is clear enough (it seems like we're referring to what was just said). I would rephrase along the lines of: "An alternative usage of this threshold function is to pass it only a histogram, not an image."